File: frontend/src/lib/klipy.ts
import { apiUrl } from "$lib/runtime-config";
export interface KlipyGif {
  id: string;
  title: string;
  urls: {
    preview: string;
    tinygif: string;
    mediumgif: string;
    gif: string;
  };
}

interface KlipyMediaFormat {
  url: string;
  width: number;
  height: number;
  size: number;
}

interface KlipyMediaSizes {
  gif?: KlipyMediaFormat;
  webp?: KlipyMediaFormat;
  jpg?: KlipyMediaFormat;
  mp4?: KlipyMediaFormat;
  webm?: KlipyMediaFormat;
}

interface KlipyFile {
  hd?: KlipyMediaSizes;
  md?: KlipyMediaSizes;
  sm?: KlipyMediaSizes;
  xs?: KlipyMediaSizes;
}

interface KlipyResponse {
  result: boolean;
  data: {
    data: Array<{
      id: number;
      slug: string;
      title: string;
      file?: KlipyFile;
      blur_preview?: string;
      type: string;
    }>;
    current_page?: number;
    per_page?: number;
    has_next?: boolean;
  };
}

// A function, not a const. Evaluated at import time this captured whatever
// the build inlined, before /config.json had been read.
// No hardcoded origin here. An unset apiUrl means this instance did not say
// where its relay is, and pointing at awful.frav.in instead would send a
// self-hoster's users - and the urls they open - to a stranger's server
// without either party knowing. Empty resolves same-origin, which 404s: the
// feature is off, and it is off HERE.
const API_URL = () => apiUrl() + "/klipy";

function normalizeGif(item: KlipyResponse["data"]["data"][0]): KlipyGif {
  const file = item.file || {};
  const sm = file.sm || file.xs || {};
  const md = file.md || file.sm || {};
  const hd = file.hd || file.md || {};
  return {
    id: String(item.id),
    title: item.title || "",
    urls: {
      preview:
        (sm as KlipyMediaSizes).webp?.url ||
        (sm as KlipyMediaSizes).jpg?.url ||
        (md as KlipyMediaSizes).webp?.url ||
        "",
      tinygif: (sm as KlipyMediaSizes).gif?.url || "",
      mediumgif: (md as KlipyMediaSizes).gif?.url || "",
      gif:
        (hd as KlipyMediaSizes).gif?.url ||
        (md as KlipyMediaSizes).gif?.url ||
        (sm as KlipyMediaSizes).gif?.url ||
        "",
    },
  };
}

export interface KlipyResult {
  gifs: KlipyGif[];
  hasMore: boolean;
}

export async function searchGifs(
  query: string,
  limit = 12,
  page = 1
): Promise<KlipyResult> {
  try {
    const params = new URLSearchParams({
      q: query,
      limit: String(limit),
      page: String(page),
    });
    const res = await fetch(`${API_URL()}/search?${params}`);
    if (!res.ok) return { gifs: [], hasMore: false };
    const data: KlipyResponse = await res.json();
    if (!data.result) return { gifs: [], hasMore: false };
    return {
      gifs: data.data.data
        .map(normalizeGif)
        .filter((g) => g.urls.tinygif || g.urls.mediumgif),
      hasMore: data.data.has_next ?? false,
    };
  } catch {
    return { gifs: [], hasMore: false };
  }
}

export async function getTrendingGifs(
  limit = 12,
  page = 1
): Promise<KlipyResult> {
  try {
    const params = new URLSearchParams({
      limit: String(limit),
      page: String(page),
    });
    const res = await fetch(`${API_URL()}/trending?${params}`);
    if (!res.ok) return { gifs: [], hasMore: false };
    const data: KlipyResponse = await res.json();
    if (!data.result) return { gifs: [], hasMore: false };
    return {
      gifs: data.data.data
        .map(normalizeGif)
        .filter((g) => g.urls.tinygif || g.urls.mediumgif),
      hasMore: data.data.has_next ?? false,
    };
  } catch {
    return { gifs: [], hasMore: false };
  }
}

File: relay/klipy.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"
)

var (
	klipyAPIBase = "https://api.klipy.com/api/v1"
	klipyAPIKey  = os.Getenv("KLIPY_API_KEY")
)

const (
	// /klipy/search and /klipy/trending are publicly routed and
	// unauthenticated (isAllowedOrigin returns true for a request with no
	// Origin header at all), so without a budget anyone with curl can spend
	// the operator's paid KLIPY_API_KEY quota. The GIF picker is chatty - a
	// debounced typeahead plus infinite scroll, and search and trending share
	// this one bucket because they share one paid quota - so the budget is
	// looser than pluginProxyRateLimit's 10/min. Shared-NAT offices sit behind
	// one apparent IP, which is the other reason not to set it tight.
	klipyRateLimit = 60
	// An upstream error page is logged for diagnosis; cap the read so a
	// misbehaving or hostile upstream cannot stream unbounded bytes into
	// memory and the log.
	klipyMaxErrorBody = 4 << 10
)

// One client with an explicit timeout for every Klipy call. http.Get uses
// http.DefaultClient, which has NO timeout at all: a hung upstream would pin
// the handler goroutine and its socket forever.
var klipyHTTPClient = &http.Client{Timeout: 10 * time.Second}

// klipyRedact strips the app key out of anything headed for the log. The key
// is a PATH segment of every upstream URL, and *url.Error prints that URL
// verbatim (net/http redacts only userinfo), so logging a raw transport error
// wrote the operator's paid key straight into the relay's logs.
func klipyRedact(s string) string {
	if klipyAPIKey == "" {
		return s
	}
	return strings.ReplaceAll(s, klipyAPIKey, "<redacted>")
}

// klipyIntParam parses a caller-supplied count as a number so it reaches the
// upstream URL as digits. limit and page were interpolated into the query
// string unescaped, which let a caller append arbitrary upstream parameters.
func klipyIntParam(raw string, def, max int) int {
	n, err := strconv.Atoi(strings.TrimSpace(raw))
	if err != nil || n < 1 {
		return def
	}
	if n > max {
		return max
	}
	return n
}

// KlipyGifResponse represents a GIF from Klipy API
type KlipyGifResponse struct {
	ID      string `json:"id"`
	URL     string `json:"url"`
	Width   int    `json:"width"`
	Height  int    `json:"height"`
	Size    int    `json:"size"`
	Preview string `json:"preview"`
	Title   string `json:"title"`
}

// KlipySearchResponse represents the search API response (use raw message to handle any structure)
type KlipySearchResponse struct {
	Result bool                   `json:"result"`
	Data   map[string]interface{} `json:"data"`
}

type KlipyMediaFormat struct {
	URL    string `json:"url"`
	Width  int    `json:"width"`
	Height int    `json:"height"`
	Size   int    `json:"size"`
}

type KlipyMediaSizes struct {
	Gif  *KlipyMediaFormat `json:"gif,omitempty"`
	Webp *KlipyMediaFormat `json:"webp,omitempty"`
	Jpg  *KlipyMediaFormat `json:"jpg,omitempty"`
	Mp4  *KlipyMediaFormat `json:"mp4,omitempty"`
	Webm *KlipyMediaFormat `json:"webm,omitempty"`
}

type KlipyFile struct {
	Hd *KlipyMediaSizes `json:"hd,omitempty"`
	Md *KlipyMediaSizes `json:"md,omitempty"`
	Sm *KlipyMediaSizes `json:"sm,omitempty"`
	Xs *KlipyMediaSizes `json:"xs,omitempty"`
}

type KlipyGifItem struct {
	ID          int        `json:"id"`
	Slug        string     `json:"slug"`
	Title       string     `json:"title"`
	File        *KlipyFile `json:"file,omitempty"`
	BlurPreview string     `json:"blur_preview,omitempty"`
	Type        string     `json:"type"`
}

type KlipyResponseData struct {
	Data        []KlipyGifItem `json:"data"`
	CurrentPage int            `json:"current_page,omitempty"`
	PerPage     int            `json:"per_page,omitempty"`
	HasNext     bool           `json:"has_next,omitempty"`
}

type KlipyAPIResponse struct {
	Result bool              `json:"result"`
	Data   KlipyResponseData `json:"data"`
}

func handleKlipySearch(w http.ResponseWriter, r *http.Request) {
	if !isAllowedOrigin(r.Header.Get("Origin")) {
		apiError(w, r, "Origin not allowed", http.StatusForbidden)
		return
	}
	if !rateAllow("klipy:"+clientIP(r), klipyRateLimit) {
		apiError(w, r, "rate limited", http.StatusTooManyRequests)
		return
	}
	if klipyAPIKey == "" {
		apiError(w, r, "KLIPY_API_KEY not configured", http.StatusServiceUnavailable)
		return
	}

	q := r.URL.Query().Get("q")
	limit := klipyIntParam(r.URL.Query().Get("limit"), 18, 50)
	page := klipyIntParam(r.URL.Query().Get("page"), 1, 1000)

	apiURL := fmt.Sprintf("%s/%s/gifs/search?q=%s&limit=%d&page=%d",
		klipyAPIBase,
		klipyAPIKey,
		url.QueryEscape(q),
		limit,
		page,
	)

	resp, err := klipyHTTPClient.Get(apiURL)
	if err != nil {
		log.Printf("[klipy] search fetch error: %s", klipyRedact(err.Error()))
		apiError(w, r, "Failed to fetch from Klipy", http.StatusInternalServerError)
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, klipyMaxErrorBody))
		log.Printf("[klipy] search API error: %d - %s", resp.StatusCode, klipyRedact(string(body)))
		apiError(w, r, "Failed to fetch from Klipy", resp.StatusCode)
		return
	}

	var result KlipyAPIResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		log.Printf("[klipy] search decode error: %s", klipyRedact(err.Error()))
		apiError(w, r, "Failed to decode Klipy response", http.StatusInternalServerError)
		return
	}

	withCors(w, r, func(w http.ResponseWriter) {
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(result)
	})
}

func handleKlipyTrending(w http.ResponseWriter, r *http.Request) {
	if !isAllowedOrigin(r.Header.Get("Origin")) {
		apiError(w, r, "Origin not allowed", http.StatusForbidden)
		return
	}
	if !rateAllow("klipy:"+clientIP(r), klipyRateLimit) {
		apiError(w, r, "rate limited", http.StatusTooManyRequests)
		return
	}
	if klipyAPIKey == "" {
		apiError(w, r, "KLIPY_API_KEY not configured", http.StatusServiceUnavailable)
		return
	}

	limit := klipyIntParam(r.URL.Query().Get("limit"), 18, 50)
	page := klipyIntParam(r.URL.Query().Get("page"), 1, 1000)

	apiURL := fmt.Sprintf("%s/%s/gifs/trending?limit=%d&page=%d",
		klipyAPIBase,
		klipyAPIKey,
		limit,
		page,
	)

	resp, err := klipyHTTPClient.Get(apiURL)
	if err != nil {
		log.Printf("[klipy] trending fetch error: %s", klipyRedact(err.Error()))
		apiError(w, r, "Failed to fetch from Klipy", http.StatusInternalServerError)
		return
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, klipyMaxErrorBody))
		log.Printf("[klipy] trending API error: %d - %s", resp.StatusCode, klipyRedact(string(body)))
		apiError(w, r, "Failed to fetch from Klipy", resp.StatusCode)
		return
	}

	var result KlipyAPIResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		log.Printf("[klipy] trending decode error: %s", klipyRedact(err.Error()))
		apiError(w, r, "Failed to decode Klipy response", http.StatusInternalServerError)
		return
	}

	withCors(w, r, func(w http.ResponseWriter) {
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(result)
	})
}
